MAINT: improve license handling, use PEP 639 license files inclusion

DimitriPapadopoulos opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue

Closes gh-17470.

#### What does this implement/fix?

Use PEP 639 instead of legacy syntax for licenses in `pyproject.toml`.

#### Additional information

Requires meson-python [0.18.0](https://mesonbuild.com/meson-python/changelog.html).

DimitriPapadopoulos committed:
MAINT: PEP 639 compliance

[skip ci]

DimitriPapadopoulos committed:
MAINT: PEP 639 compliance

[skip ci]

lucascolley reviewed (commented):
do we need to change anything with regards to bundled licenses here?

DimitriPapadopoulos commented on pyproject.toml:
```diff
@@ -34,9 +34,8 @@ requires = [
 name = "scipy"
 version = "1.17.0.dev0"
-# TODO: add `license-files` once PEP 639 is accepted (see meson-python#88)
-#       at that point, no longer include them in `py3.install_sources()`
```
I'm afraid I don't really understand the comment. I'll look into its history and try to understand what it means.

By the way, there are a few issues with licencing:
1. Some paths to licence files of bundled licence files were incorrect, I fixed them.
2. Consistently name licences using the SPDX Short Identifier.

DimitriPapadopoulos replied:
> do we need to change anything with regards to bundled licenses here?

I could add `LICENSES_bundled.txt` to `license-files`.

tylerjereddy replied:
My interpretation of that comment is that you actually need to adjust the `meson` build process to stop packing in license files since the PEP should handle that. I'll paste a partial diff below the fold, showing a potential change for one of the files (there are others...). That said, we may want to wait for Ralf to confirm that we are "green light" for doing this.. IIRC there was some question about `pip` version support upstream in NumPy, but I guess that actually wasn't a problem in the end.

<details>

```diff
--- a/scipy/integrate/meson.build
+++ b/scipy/integrate/meson.build
@@ -143,7 +143,6 @@ py3.install_sources([
     'odepack.py',
     'quadpack.py',
     'vode.py',
-    'src/LICENSE_DOP'
   ],
```

</details>

rgommers reviewed (changes requested):
Thanks @DimitriPapadopoulos for tackling this. We are indeed unblocked now to do this. Can you please do the following:
- Open a separate PR for the fixes to `LICENSES_bundled.txt`, since those changes can be backported to the 1.16.x branch.
- Rebase this PR on top of that one after it gets merged and in it, in addition to addressing my inline comments:
    - Delete `LICENSES_bundled.txt`
    - Populate `license` and `license_files` in the top-level `meson.build` file of this repo and the different `subprojects/*/meson.build` files
    - Move `license` and `license_files` to `dynamic=` in `pyproject.toml`
    - Verify that if you build a wheel, it contains all the license files in `.dist-info/` and none anymore under `scipy/`

In addition, we'll have to possibly change something to the license handling in `tools/wheels/` but I haven't figured out that plumbing yet - I'm looking at that for NumPy so we can copy it from there later.

rgommers commented:
@DimitriPapadopoulos it'd be fine to include `[skip ci]` in your commit messages. You're triggering a large battery of CI jobs if you push commits one by one.

DimitriPapadopoulos committed:
MAINT: Handle licensing of bundled software with Meson

In the top-level `meson.build` use the `license` and `license_files`
arguments of `project()`. Make `license` and `license-files` dynamic
in `pyproject.toml`.

All license files listed in `LICENSES_bundled.txt` have been moved
to the `license_files` argument. Source files containing their own
licensing information cannot be formally listed here, but appear as
comments. Comments alongside each entry show the SPDX identifier of
each license, they are all collected in the `license` argument.

Found a few unlisted license files in the process.

Some entries from `LICENSES_bundled.txt` cannot easily be adapted,
I have left them in there pending further discussion.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Handle licensing of bundled software with Meson

In the top-level `meson.build` use the `license` and `license_files`
arguments of `project()`. Make `license` and `license-files` dynamic
in `pyproject.toml`.

All license files listed in `LICENSES_bundled.txt` have been moved
to the `license_files` argument. Source files containing their own
licensing information cannot be formally listed here, but appear as
comments. Comments alongside each entry show the SPDX identifier of
each license, they are all collected in the `license` argument.

Found a few unlisted license files in the process.

Some entries from `LICENSES_bundled.txt` cannot easily be adapted,
I have left them in there pending further discussion.

[skip ci]

DimitriPapadopoulos commented on meson.build:
```diff
@@ -2,7 +2,6 @@ project(
   'scipy',
   'c', 'cpp', 'cython',
   version: run_command(['tools/gitversion.py'], check: true).stdout().strip(),
-  license: 'BSD-3',
```
It remains unclear to me in which `meson.build` files I should add/modify [license](https://mesonbuild.com/Reference-manual_functions.html#project_license) and [license_files](https://mesonbuild.com/Reference-manual_functions.html#project_license_files). The top-level `meson.build` file only?  Bundled code with its own subdirectory and `meson.build` file could be handled separately, inside its own  `meson.build` file.

Then there's the non-standard licence of Qhull. I'll have to read on how to handle such licences in SPDX expressions.

lucascolley replied:
I think you want `LicenseRef-...`, https://spdx.github.io/spdx-spec/v2.3/SPDX-license-expressions/

DimitriPapadopoulos replied:
I've created a `LicenseRef-Qhull` licence identifier. However, I don't understand if I can link to the licence file (other than in comments) to clarify the meaning of `LicenseRef-Qhull` and improve machine-readability.

DimitriPapadopoulos replied:
By the way, I just noticed this clause of the Qhull licence:
> ```
> 3. If you modify Qhull, you must include a notice giving the
>    name of the person performing the modification, the date of
>    modification, and the reason for such modification.
> ```
I'm not sure this currently is or can be handled well...

rgommers commented:
Now that gh-23452 is merged, can you rebase? And delete `LICENSES_bundled.txt` completely?

DimitriPapadopoulos committed:
MAINT: Add license files from `subprojects`

[skip ci]

DimitriPapadopoulos committed:
MAINT: Add license files from `subprojects`

[skip ci]

DimitriPapadopoulos committed:
MAINT: Special case for LAPJVsp licensing

* Create a subdirectory for LAPJVsp and add a dummy `__init__.py`.
* Move `_matching.pyx` to that subdirectory.
* Add a `LICENSE` file with content copied from `LICENSES_bundled.txt`.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Special case for LAPJVsp licensing

* Create a subdirectory for LAPJVsp and add a dummy `__init__.py`.
* Move `_matching.pyx` to that subdirectory.
* Add a `LICENSE` file with content copied from `LICENSES_bundled.txt`.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Special case for pybind11 licensing

Add the URL of the pybind11 web site as a comment.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Special case for pybind11 licensing

Add the URL of the pybind11 web site as a comment.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Get rid of LICENSES_bundled.txt at last

[skip ci]

DimitriPapadopoulos committed:
MAINT: Get rid of LICENSES_bundled.txt at last

[skip ci]

DimitriPapadopoulos commented:
For LAPJVsp, I have created a specific `lap_sparse` subdirectory to put the code and license file together. Happy to change the name or the method as explained in https://github.com/scipy/scipy/pull/12541#issuecomment-3200365888.

DimitriPapadopoulos committed:
Revert "MAINT: Special case for pybind11 licensing"

This reverts commit 97132ea679561dfc1d4636fa3c8b2d3136d4616c.

[skip ci]

DimitriPapadopoulos committed:
Revert "MAINT: Special case for pybind11 licensing"

This reverts commit 97132ea679561dfc1d4636fa3c8b2d3136d4616c.

[skip ci]

DimitriPapadopoulos committed:
Revert "MAINT: Special case for LAPJVsp licensing"

This reverts commit fd2d3c901dfdf0f1bc5e6f75535c880d085c4891.

[skip ci]

DimitriPapadopoulos committed:
Revert "MAINT: Special case for LAPJVsp licensing"

This reverts commit fd2d3c901dfdf0f1bc5e6f75535c880d085c4891.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Better solution for LAPJVsp licensing

[skip ci]

DimitriPapadopoulos committed:
MAINT: Better solution for LAPJVsp licensing

[skip ci]

DimitriPapadopoulos committed:
MAINT: Add pybind11, NumPy, Pythran LICENSE files

Add them under a LICENSES top-level subdirectory.

Copy the LICENSE files verbatim from the respective GitHub repositories.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Add pybind11, NumPy, Pythran LICENSE files

Add them under a LICENSES top-level subdirectory.

Copy the LICENSE files verbatim from the respective GitHub repositories.

[skip ci]

rgommers reviewed (commented):
This is starting to look good. There is a nontrivial problem left though: license content in `.pyx`, `.c`, `.h` etc. files doesn't end up in wheels, it gets discarded in the compilation stage. 

License content in `.py` files does get included, but it isn't very discoverable. I'm inclined to solve both problems by copying/moving all license content as separate files under `LICENSES`. That does have the downside that we have two copies for some separate files. But the upside that it's all very easy to inspect then. 

@DimitriPapadopoulos what do you think about that?

DimitriPapadopoulos committed:
MAINT: files covered by the NumPy license

[skip ci]

DimitriPapadopoulos committed:
MAINT: files covered by the NumPy license

[skip ci]

DimitriPapadopoulos commented:
The end result will be very close to the initial `LICENSES_distributed.txt` file, just:
1. split into a singe file for each 3rd-party piece of code,
2. without information about the actual piece of code covered by each file.

I'm starting to think the initial `LICENSES_distributed.txt` file wasn't that bad. I don't really have an opinion concerning point 1. I'm more concerned about point 2, the now missing information might be useful for future maintenance.

rgommers commented:
> without information about the actual piece of code covered by each file.

We can add a `LICENSES/README` to point out things that are relevant, if that makes sense.

There's definitely no silver bullet for this - licensing is complex - but after this last change it will be much more complete, as well as machine-parseable.

DimitriPapadopoulos committed:
MAINT: rework licensing for getLebedevSphere

* Get the original `license.txt` from the Matlab code ZIP file.
* Remove licensing details from the Python source code.
* Add some additional copyright/licensing stuff from the initial Matlab code.

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for getLebedevSphere

* Get the original `license.txt` from the Matlab code ZIP file.
* Remove licensing details from the Python source code.
* Add some additional copyright/licensing stuff from the initial Matlab code.

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for cctbx

* Get the original `LICENSE_2_0.txt` from:
  https://cci.lbl.gov/cctbx_sources/boost_adaptbx/

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for cctbx

* Get the original `LICENSE_2_0.txt` from:
  https://cci.lbl.gov/cctbx_sources/boost_adaptbx/

[skip ci]

DimitriPapadopoulos committed:
MAINT: Add a README file under LICENSES

It explains which piece of code the LICENSE files applies to.

[skip ci]

DimitriPapadopoulos committed:
MAINT: Add a README file under LICENSES

It explains which piece of code the LICENSE files applies to.

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for LAP_sparse/LAPJVS/lapjvps

* Get the licensing terms from:
  scipy/sparse/csgraph/_matching.pyx

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for LAP_sparse/LAPJVS/lapjvps

* Get the licensing terms from:
  scipy/sparse/csgraph/_matching.pyx

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for L-BFGS-B python wrapper

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for L-BFGS-B python wrapper

[skip ci]

DimitriPapadopoulos committed:
MAINT: remove references to non-existent files

[skip ci]

DimitriPapadopoulos committed:
MAINT: remove references to non-existent files

[skip ci]

DimitriPapadopoulos committed:
MAINT: clean up

[skip ci]

DimitriPapadopoulos committed:
MAINT: clean up

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for L-BFGS-B

* Get the original `License.txt` from L-BFGS-B 3.0:
  https://users.iems.northwestern.edu/~nocedal/lbfgsb.html

[skip ci]

DimitriPapadopoulos committed:
MAINT: rework licensing for L-BFGS-B

* Get the original `License.txt` from L-BFGS-B 3.0:
  https://users.iems.northwestern.edu/~nocedal/lbfgsb.html

[skip ci]

DimitriPapadopoulos committed:
MAINT: Add new license files to distributions

[skip ci]

DimitriPapadopoulos committed:
MAINT: Add new license files to distributions

[skip ci]

rgommers commented:
> Examples of files that may require their own licensing terms (or not)

Yes, I'm afraid that a good fraction of those do indeed need to be handled. And agreed that that is best done in a separate PR.

rgommers committed:
BUG: fix issues in top-level meson.build files

[ci skip]

rgommers committed:
BUG: fix issues in top-level meson.build files

[ci skip]

rgommers commented:
I pushed a fix for some issues in this PR. Remaining is:
```
../meson.build:1:0: ERROR: Sandbox violation: Tried to grab file LICENSE from a nested subproject.
```
That is indeed not allowed. Will check if that's handled correctly by `meson`/`meson-python` before trying to fix.

tylerjereddy commented:
@rgommers I guess this hasn't been touched in about half a year--do we want to try to get this in in around ~10-11 days before branching, or aim for just after?

rgommers commented:
It's not easy to resolve, bumping the milestone

DimitriPapadopoulos commented:
Rebased.

rgommers commented:
I refreshed my memory partially. The use of `dynamic` in `pyproject.toml` is a bit unfortunate, but seemed necessary. I think [PEP 808](https://peps.python.org/pep-0808/) is about to be approved, and that would allow us to move the static list of license file into `pyproject.toml` - unless the move to `dynamic` was to avoid the sandbox violation. That needs testing again. NumPy does have `license-files` in `pyproject.toml`, which is nicer.

> In addition, we'll have to possibly change something to the license handling in `tools/wheels/` but I haven't figured out that plumbing yet - I'm looking at that for NumPy so we can copy it from there later.

xref https://github.com/numpy/numpy-release/issues/11
